domain-models-runtime/src/main/java/org/folio/rest/persist/Conn.java

```diff
@@ -297,4 +306,133 @@ static String durationMsg(String descriptionKey, String sql, long startNanoTime)
+          JsonObject jsonObject = new JsonObject(json);
+          String id = jsonObject.getString("id");
+          list.add(Tuple.of(
+              id == null ? UUID.randomUUID() : UUID.fromString(id),
+              jsonObject));
+        }
+      }
+      return saveBatchInternal(upsert, table, list);
+    } catch (Exception e) {
+      log.error(e.getMessage(), e);
+      return Future.failedFuture(e);
+    }
+  }
+
+  private RowSet<Row> emptyRowSetOfId() {
+    return new LocalRowSet(0).withColumns(Collections.singletonList("id"));
+  }
+
+  /**
+   * @throws Throwable the caller needs to catch and convert into a failed Future
+   */
+  private Future<RowSet<Row>> saveBatchInternal(boolean upsert, String table, List<Tuple> batch) {
+
+    if (batch.isEmpty()) {
+      // vertx-pg-client fails with "Can not execute batch query with 0 sets of batch parameters."
+      return Future.succeededFuture(emptyRowSetOfId());
+    }
+    long start = log.isDebugEnabled() ? System.nanoTime() : 0;
+    log.info("starting: saveBatch size=" + batch.size());
+    String sql = "INSERT INTO " + postgresClient.getSchemaName() + "." + table
+        + " (id, jsonb) VALUES ($1, $2)"
+        + (upsert ? " ON CONFLICT (id) DO UPDATE SET jsonb = EXCLUDED.jsonb" : "")
+        + " RETURNING id";
+    return pgConnection.preparedQuery(sql).executeBatch(batch)
+    .map(rowSet -> {
+      log.debug(() -> durationMsg("saveBatch", table, start));
+      if (rowSet == null) {
+        return emptyRowSetOfId();
+      }
+      return rowSet;
+    })
+    .onFailure(e -> {
+      log.error("saveBatch size=" + batch.size() + " " + e.getMessage(), e);
+      log.debug(() -> durationMsg("saveBatchFailed", table, start));
+    });
+  }
+
+  /**
+   * Insert the entities into table using a single INSERT statement.
+   * @param table  destination table to insert into
+   * @param entities  each array element is a String with the content for the JSONB field of table; if id is missing a random id is generated
+   * @return one result row per inserted row, containing the id field
+   */
+  public Future<RowSet<Row>> saveBatch(String table, JsonArray entities) {
+    return saveBatch(false, table, entities);
+  }
+
+  /**
+   * Upsert the entities into table using a single INSERT statement.
+   * @param table  destination table to insert into
+   * @param entities  each array element is a String with the content for the JSONB field of table; if id is missing a random id is generated
+   * @return one result row per inserted row, containing the id field
+   */
+  public Future<RowSet<Row>> upsertBatch(String table, JsonArray entities) {
+    return saveBatch(true, table, entities);
+  }
+
+  <T> Future<RowSet<Row>> saveBatch(boolean upsert, String table, List<T> entities) {
+
+    try {
+      List<Tuple> batch = new ArrayList<>();
+      if (entities == null || entities.isEmpty()) {
+        return Future.succeededFuture(emptyRowSetOfId());
+      }
```

Comment:
This does not need to be in the try block

Reply:
entities.isEmpty() may throw an exception because it may be any custom List implementation. Putting it into the try-catch block ensures that the exception is wrapped into a failed Future.

Reply:
Fine. Can you switch the order of `batch` declaration and `entities` check? There is no reason to declare the `batch` if `entities` check could fail.